README.md
![logo of Walder](logo/logo.png)

Walder offers an easy way to set up a website or Web API on top of decentralized knowledge graphs.
Knowledge graphs incorporate data together with the meaning of that data.
This makes it possible to combine data from multiple knowledge graphs,
even if different, independent parties maintain or host them.
Knowledge graphs can be hosted via 
[Solid](https://solidproject.org/) PODs, 
[SPARQL](https://www.w3.org/TR/sparql11-query/) endpoints, 
[Triple Pattern Fragments](https://linkeddatafragments.org/) interfaces, 
[RDF](https://www.w3.org/TR/rdf11-concepts/) files, 
and so on.
Using content negotiation, 
Walder makes the data in these knowledge graphs available to clients 
via HTML, RDF, and JSON-LD.
Users define in a configuration file which data Walder uses and 
how it processes this data.
Find out which APIs are built with Walder [here](#built-with-walder).

**Table of contents**

- [Installation](#installation)
- [Usage](#usage)
  - [CLI](#cli)
  - [Library](#library)
  - [Config file structure](#config-file-structure)
    - [Resources](#resources)
    - [Path entry](#path-entry)
  - [Example](#example)
  - [Options for GraphQL-LD queries](#options-for-graphql-ld-queries)
  - [Multiple config files](#multiple-config-files)
  - [Content negotiation](#content-negotiation)
    - [RDF](#rdf)
  - [HTML templates](#html-templates)
    - [Accessing query results in view templates](#accessing-query-results-in-view-templates)
    - [Using layouts in view templates](#using-layouts-in-view-templates)
    - [Accessing front-matter metadata in view templates and layout templates](#accessing-front-matter-metadata-in-view-templates-and-layout-templates)
- [Input validation](#input-validation)
- [Error handling](#error-handling)
  - [Errors](#errors)
    - [Global](#global)
    - [Pipe modules](#pipe-modules)
    - [GraphQL-LD](#graphql-ld)
  - [Example](#example-1)
- [Developing your website](#developing-your-website)
- [Dependencies](#dependencies)
- [Tests](#tests)
- [Built with Walder](#built-with-walder)
- [License](#license)

## Installation

Install Walder globally via `yarn global add walder`.

For development, follow these steps:

1. Clone the repository.
2. Install dependencies via `yarn install`.
3. Install Walder globally via `$ yarn global add file:$(pwd)`.

## Usage

Walder is available as a [CLI](#cli) and JavaScript [library](#library).

### CLI

```bash
Usage: walder [options]

Options:
  -v, --version              output the version number
  -c, --config <configFile>  YAML configuration file input
  -p, --port [portNumber]    server port number (default: 3000)
  -l, --log [level]          enable logging and set logging level (one of [error, warn, info, verbose, debug]) (default: "info")
  --no-cache                 disable Comunica default caching
  --lenient                  turn Comunica errors on invalid data into warnings
  -h, --help                 output usage information
```

### Library

```js
// From the root directory
const Walder = require('.');

const configFilePath = 'path/to/configfile';
const port = 9000; // Defaults to 3000 
const logging = 'info'; // Defaults to 'info' 
const cache = false;  // Defaults to true
const lenient = true; // Defaults to false 

const walder = new Walder(configFilePath, {port, logging, cache, lenient, cwd});

walder.activate();    // Starts the server
walder.deactivate();  // Stops the server
```

### Config file structure

You write a config file in YAML following the [OpenAPI 3.0 specification](https://swagger.io/docs/specification/basic-structure/).
The config file must have the following structure:

```yaml
openapi: 3.0.2
info:  # OpenAPI metadata
  title: 'Example site'
  version: 0.1.0
x-walder-resources:  # Directories used by Walder - OPTIONAL
  root:  # Path to the root folder of the directories used by Walder (absolute or relative to the directory containing the config file) - OPTIONAL (default: .)
  views:  # Path to directory containing view template files (absolute or relative to the root folder) - OPTIONAL (default: views)
  pipe-modules:  # Path to directory containing local pipe modules (absolute or relative to the root folder) - OPTIONAL (default: pipe-modules)
  public:  # Path to directory containing all files that should be available statically (e.g. stylesheets) (absolute or relative to the root folder) - OPTIONAL (default: public)
  layouts: # Path to directory containing all layout template files that can be used by view template files (absolute or relative to the root folder) - OPTIONAL (default: layouts)
x-walder-datasources:  # Default list of data sources
  - ...  # E.g. link to SPARQL endpoint or a GraphQL-LD query
paths:  # List of path entries
  path-entry-1:
    ...
  path-entry-2:
    ...
x-walder-errors: # Default error page views - status codes with files containing the HTML view template (absolute path or relative to the views directory)
  404: ...
  500: ...
  ...
```

#### Resources

The `x-walder-resources` key of the config file contains paths to directories used by Walder.
This key and it's values are optional. 
If a user does not give paths, 
Walder uses the following default values relative to the directory of the config file.

```yaml
root: .
views: views
pipe-modules: pipe-modules
public: public
layouts: layouts
```

To prevent Walder from making the wrong files public, 
when a user does not give a path to the `public` field, 
Walder creates a new directory `public` if it does not find this directory in the current working directory and uses that one.

#### Path entry

A path entry defines a route and has the following structure:

```yaml
path:  # The path linked to this query
  request:  # The HTTP request type (GET, POST, etc.)
    summary: ...  # Short description
     parameters:  # Path variables/Query parameters
        - in: ...  # 'path' or 'query'
          name: ...  # Name of the parameter
          schema:
            type: ... # Type of the parameter
          description: ...  # Description of the parameter
    x-walder-query:
      graphql-query: ...  # One or more GraphQL queries
      json-ld-context: ...  # The JSON-LD corresponding to the GraphQL query
      sparql-query: ... # One or more SPARQL queries
      json-ld-frame: ... # A JSON-LD frame that should be applied to the result of a SPARQL query 
      options: # Global options that will be applied to all the graphql-queries of this path (OPTIONAL)
      datasources:  # Query specific datasources (OPTIONAL)
        additional: ...  # Boolean stating that the following datasources are meant to be used on top of the default ones
        sources:  # List of query specific datasources
          - ...  # E.g. link to SPARQL endpoint
    x-walder-postprocessing:  # The (list of) pipe modules used for postprocessing
      module-id:  # Identifier of the pipe module
        source: ...  # Path leading to source code of the pipe module (absolute path or relative to the pipe-modules directory)
        parameters: # the parameters for the pipe module (OPTIONAL)
          - _data # (DEFAULT) this gives all the data, but all paths in the data object are supported (e.g. _data.0.employee)
          - ... # Additional parameters if your function supports those (OPTIONAL)
    responses:  # Status codes with files containing the HTML view template (absolute path or relative to the views directory)
      200: ...  # (REQUIRED)
      500: ...  # (OPTIONAL)
```

### Example

The following command starts a server on port 3000 (default) using an example config file.

`$ walder -c example/config.yaml `

This will start a server on `localhost:3000` with the following routes:

* <http://localhost:3000/bradpitt-directors> - Returns a list of directors of movies starring Brad Pitt.
* <http://localhost:3000/music/{musician}> - Returns a list of songs by a given musician.
For example, <http://localhost:3000/music/Lady%20Gaga> returns a list of songs by Lady Gaga.
* <http://localhost:3000/artist/{artist}?writer={name}> - Returns a list of a given artist's songs 
selecting those written by a specific person identified by name. 
For example, <http://localhost:3000/artist/David%20Bowie?writer=John%20Lennon> returns a list of 
a given artist's songs written by a specific person.
* <http://localhost:3000/music/{artist}/postprocessed> - Returns a list of songs by a given artist 
that have 'star' in the title, using pipe modules.
For example, <http://localhost:3000/music/David%20Bowie/postprocessed> returns a list of songs by David Bowie that 
have 'star' in the title.

### Options for GraphQL-LD queries

In the path entry above, 
the user defined `options` as a global (optional) identifier that Walder uses for every query of that path.
We have two options where we can choose from: 
`sort` and `remove-duplicates`. 
With given syntax:

```yaml
options:
  sort: # Enable sorting on the data (OPTIONAL)
    object: # JSONPath to the object you want to sort for
    selectors: # The values inside the object over which you want to sort
      - ... # The default option when you want ascending order, just give the value (JSONPath notation supported for further nesting)
      - value: ...  # When you want descending order, specify the value/order (JSONPath notation supported for further nesting)
        order: desc
  remove-duplicates: # Enable the removal of duplicates of the data (OPTIONAL)
    object: ... # The JSONPath tot the object that you want to compare
    value: ... # The value that has to be compared to determine whether it's duplicate (JSONPath notation is also supported for further nesting)
```

If you do not want `options` to be global for the whole path, 
you can define `options` per query.

```yaml
path:  # The path linked to this query
  request:  # The HTTP request type (GET, POST, etc.)
    summary: ...  # Short description
     parameters:  # Path variables/Query parameters
        - in: ...  # 'path' or 'query'
          name: ...  # Name of the parameter
          schema:
            type: ... # Type of the parameter
          description: ...  # Description of the parameter
    x-walder-query:
      graphql-query: ...  # One or more GraphQL queries
        name:
          query: ... # The GraphQL query
          options: # options that will be applied only to this specific graphql-query (OPTIONAL)
...
```

The following command starts a server using [this](example/config-sorting-duplicates.yaml) config file.

`$ walder -c example/config-sorting-duplicates.yaml`

This will start a server on `localhost:3000` with the following routes:

* <http://localhost:3000/music/{musician}/sorted> - 
Returns a list of songs with their artists written by a given musician. 
Walder sorts the songs in descending order by song name.
For example, <http://localhost:3000/music/Marcella%20Detroit/sorted> returns a list of such songs written by Marcella Detroit.
* <http://localhost:3000/music/{musician}/no_duplicates> - 
Returns a list of songs with their artists written by a given musician.
Walder removes duplicate song names from the list.
For example, <http://localhost:3000/music/Marcella%20Detroit/no_duplicates> returns a list of such songs by John Marcella Detroit.
* <http://localhost:3000/movies/{musician}/everything_together> - 
Returns a list of songs with their artists written by a given musician.
Walder sorts the songs in descending order by song name and removes duplicate song names.
For example, <http://localhost:3000/music/Marcella%20Detroit/everything_together> returns a list of such songs by Marcella Detroit.
* <http://localhost:3000/artist/{artist}> - 
Returns a list of song names and movie URLs for a given artist. 
Walder removes duplicate songs based on their names and 
sorts movies in descending order by their URLs.
For example, <http://localhost:3000/artist/David%20Bowie> returns a list of such songs and movies.

### Multiple config files

You can split a config file in multiple files, 
using the `$ref` keyword. 
We follow the [OpenAPI 3.0 spec](https://swagger.io/docs/specification/using-ref/) 
that explains how to use the referencing.

When first referenced you need to use the path beginning from the directory of the config file, 
but if the referenced file has references itself, 
it can use paths relative to its own location, as shown below.

The actual config file referencing its paths
```yaml
openapi: 3.0.2
info:
  title: 'Example site'
  version: 0.1.0
x-walder-resources:
  path: ./
  views: views
  pipe-modules: pipeModules
  public: public
x-walder-datasources:
  - http://fragments.dbpedia.org/2016-04/en
paths:
  /music/{musician}:
    $ref: './paths/music_musician.yaml'
  /movies/{actor}:
    $ref: './paths/movies_actor.yaml'
x-walder-errors:
  404:
    description: page not found error
    x-walder-input-text/html: error404.html
  500:
    description: internal server error
    x-walder-input-text/html: error500.html
```

Below you see `./example/paths/movies_actor.yaml` with reference with path relative to its own location

```yaml
get:
  summary: Returns a list of the all movies the given actor stars in
  parameters:
    - in: path
      name: actor
      required: true
      schema:
        type: string
      description: The target actor
  x-walder-query:
    $ref: '../walderQueryInfo/movies_actor_info.yaml'
  responses:
    200:
      description: list of movies
      x-walder-input-text/html: movies.pug

```

### Content negotiation

Using content negotiation, 
Walder makes the following output formats available:

* 'text/html'
* 'application/ld+json'
* 'text/turtle'
* 'application/n-triples'
* 'application/n-quads'

#### RDF

Walder uses [graphql-ld-comunica](https://www.npmjs.com/package/graphql-ld-comunica) 
to execute the GraphQL queries
and [@comunica/query-sparql](https://github.com/comunica/comunica/tree/master/engines/query-sparql) 
to execute SPARQL queries.
The result of a GraphQL-LD query is a JSON data. 
Walder first converts it into JSON-LD. 
This enables conversion to other RDF formats during content negotiation.
The result of a SPARQL query is an array of quads.
If a JSON-LD frame is specified, the quads are converted to JSON-LD.
Due to the importance of content negotiation,
only CONSTRUCT queries are supported.

### HTML templates

Walder uses [consolidate](https://www.npmjs.com/package/consolidate) 
to automatically retrieve the corresponding engine for a given template. 
This means that the [supported template engines](https://www.npmjs.com/package/consolidate#supported-template-engines) 
are dependent on consolidate.

You can use different template engines for different routes, e.g., 
[pug](https://pugjs.org/api/getting-started.html) renders one route's HTML, 
while [handlebars](https://handlebarsjs.com/) renders another route's HTML. 
Walder does this all by looking at the file extension of the given template.

Templates can be used in views as well as in layouts. So we'll name them **view templates** and **layout templates** in order to distinguish.

#### Accessing query results in view templates

The results of the queries, specified in the configuration file for a route, 
are available for rendering in view templates as data.
- If the route only has a single query, the data contains an object named `data`.
- If the route has multiple queries, 
each query's result is available in the data as a separate object 
whose name equals the query name in the configuration file.

In the case of a GraphQL-LD query,
each object will be an array, 
unless the query was [singularized](https://github.com/rubensworks/graphql-to-sparql.js#singularizing-everything).
[songs.handlebars](example/views/songs.handlebars) is an example of the consumption of the result
of the single query in the route `/music/{musician}` in [this configuration file](example/config.yaml).
[songs_movies.handlebars](example/views/songs_movies.handlebars) is an example of the consumption of the results
of the two queries in the route `/artist/{artist}` in [this configuration file](example/config-multiple-queries.yaml).

In the case of a SPARQL query, 
each object is an array of quads if no JSON-LD frame is specified.
Else it will be a JSON-LD object.

#### Using layouts in view templates

Using layouts is a great way to avoid repetition in route-specific view templates.
Reusable HTML structures such as headers, footers, navigation bars and other contents, meant to appear in multiple routes,
are preferable specified in *layout files*.

A layout template file can be specified in a view template file, 
by means of [front-matter](https://github.com/jxson/front-matter) metadata field `layout`.
It should contain a filename, available at the `layouts` location defined in the configuration file.
It may contain a relative path in front of the filename.

Example view template file specifying a layout:
```
---
layout: my-layout.pug
---
// view template continues here
```

Walder puts the inner HTML contents generated from the view template file into the data forwarded to 
the layout template file as an object named `content`.

The layout template file is yet another template. 
It usually expands these inner HTML contents at the position of its choice.

A simple *pug* example (mind the `!{content}`):
```
doctype html
html(lang="en")
    head
        title I'm based on a layout
    body !{content}
```

#### Accessing front-matter metadata in view templates and layout templates

In addition to query results, Walder adds [front-matter](https://github.com/jxson/front-matter) metadata,
specified in view templates, as additional attributes to the data.

Each additional attribute's name is equal to the metadata field name provided.
The following metadata field are reserved: `layout`, `content`, `data`, and 
the names assigned to queries in routes having multiple queries (see above).

These attributes are available to the view template and to the layout template it refers to, if any.

Example view template file specifying a front-matter metadata field and reading that field (mind the `#{a1}`):
```
--
a1: Value for FrontMatter attribute a1!
---

doctype html
html(lang="en")
    body
        main a1: #{a1}
```

Example view template file, specifying a layout template and another front-matter metadata field:
```
---
layout: layout-fm.pug
a2: Value for FrontMatter attribute a2!
---

main Lorem ipsum
```

Example corresponding layout template (layout-fm.pug) reading that field (mind the `#{a2}`):
```
doctype html
html(lang="en")
    head
        if a2
            title #{a2}

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

## [4.1.4] - 2023-11-28

### Fixed
- Invalidate cache of Comunica engine
- Passing cache option to queries that return data sources for other queries

## [4.1.3] - 2023-07-18

### Updated
- Lock file to fix bug in Comunica

## [4.1.2] - 2023-05-09

### Added
- Make query results available in layouts

## [4.1.1] - 2023-03-31

### Changed
- Update Comunica deps

## [4.1.0] - 2023-03-30

### Added
- Clarify GraphQL-LD error when certificate fails (see [issue 80](https://github.com/KNowledgeOnWebScale/walder/issues/80))
- Add query results as JSON-LD to HTML output (see [issue 125](https://github.com/KNowledgeOnWebScale/walder/issues/125))
- Support for SPARQL queries

### Changed
- Use `namedNode` in `graphql-ld-handler.js`.

### Fixed
- Query for data sources within a path.

## [4.0.0] - 2022-05-02

### Fixed
- Keep named nodes when returning RDF results of query (see [issue 120](https://github.com/KNowledgeOnWebScale/walder/issues/120))
- Blank node is returned when IRI is expected when requesting Turtle (see [issue 116](https://github.com/KNowledgeOnWebScale/walder/issues/116))

### Changed
- Static IDs on the first level of a query are included in query results.

## [3.0.0] - 2022-04-08

### Changed
- Handling of query results in pipe modules

### Fixed 
- GraphQLLDInfoValidator not strict enough when validating variables (see [issue 55](https://github.com/KNowledgeOnWebScale/walder/issues/55))

## [2.2.2] - 2021-09-29

### Added
- Support for Markdown footnotes

### Changed
- Error handling in `RequestHandler`

## [2.2.1] - 2021-08-26

### Added
- Support async pipe modules (see [issue 112](https://github.com/KNowledgeOnWebScale/walder/issues/112))
- Support minimum/maximum of parameters (see [issue 24](https://github.com/KNowledgeOnWebScale/walder/issues/24))

### Changed
- Switch to Github actions for CI  (see [issue 102](https://github.com/KNowledgeOnWebScale/walder/issues/102))
- Update `graphql-ld` dep

### Fixed
- Config-file-parser shouldn't export async function (see [issue 54](https://github.com/KNowledgeOnWebScale/walder/issues/54))
- Enable logging of express again (see [issue 43](https://github.com/KNowledgeOnWebScale/walder/issues/43))
- Throws error when converting empty data to Turtle

## [2.2.0] - 2021-08-04

### Added
- Template files are loaded once at the start and then cached with their front-matter (see [issue 87](https://github.com/KNowledgeOnWebScale/walder/issues/87))
- .njk is interpreted as a nunjuck template (see [issue 75](https://github.com/KNowledgeOnWebScale/walder/issues/75))
- Error when integer in graphql query variable can't be parsed (see [issue 84](https://github.com/KNowledgeOnWebScale/walder/issues/84))
- File validation error message now reflects when error originates from layout file
- Error when integer in graphql query variable can't be parsed (see [issue 84](https://github.com/KNowledgeOnWebScale/walder/issues/84))

### Fixed
- html-convertor can't convert a template using a layout that in turn extends another layout
- Class RdfConverter does not obey Convertor interface (see [issue 58](https://github.com/KNowledgeOnWebScale/walder/issues/58))
- Fix memory leak  (see [issue 83](https://github.com/KNowledgeOnWebScale/walder/issues/83))
- Incorrect error when frontmatter is invalid (see [issue 105](https://github.com/KNowledgeOnWebScale/walder/issues/105))
- Tests passing wrongly (see [issue 60](https://github.com/KNowledgeOnWebScale/walder/issues/60))

### Changed
- Update Comunica deps  (see [issue 98](https://github.com/KNowledgeOnWebScale/walder/issues/98))

## [2.1.0] - 2021-04-22

### Added
- Added tests for default error pages `*.pug`, `*.handlebars`, `*.md` (see [issue 49](https://github.com/KNowledgeOnWebScale/walder/issues/49))
- Documented the consumption of query results in templates (see [issue 62](https://github.com/KNowledgeOnWebScale/walder/issues/62))
- Documented how to use layouts (see [issue 63](https://github.com/KNowledgeOnWebScale/walder/issues/63))
- Supporting comunica lenient querying (see [issue 68](https://github.com/KNowledgeOnWebScale/walder/issues/68))
- FrontMatter metadata available in layout templates (see [issue 72](https://github.com/KNowledgeOnWebScale/walder/issues/72))
- Logging to HtmlConverter
- Default support for Nunjucks templates

### Changed
- Update dependencies
- Methods of HtmlConverter are non-static anymore
- Read file in HtmlConverter sync
- Update deps: `fs-extra`

### Fixed
- Filter query/path parameters before substituting variables (see [issue 36](https://github.com/KNowledgeOnWebScale/walder/issues/36))
- 404 error on /favicon.ico in examples (see [issue 31](https://github.com/KNowledgeOnWebScale/walder/issues/31))
- Deprecation warning for x-walder-resources.path is wrong and does not appear (see [issue 52](https://github.com/KNowledgeOnWebScale/walder/issues/52))
- Better logging when file is not found for route + detection at activation time + idem for default error pages (see [issue 48](https://github.com/KNowledgeOnWebScale/walder/issues/48))
- Dependency update to support handling of JSON-LD referring to http://schema.org (see [issue 64](https://github.com/KNowledgeOnWebScale/walder/issues/64))
- If Walder is running and the data changes, the data isn't refetched and the query doesn't re-run, even with --no-cache option'  (see [issue 23](https://github.com/KNowledgeOnWebScale/walder/issues/23))
- Multiple path parameters are not supported  (see [issue 81](https://github.com/KNowledgeOnWebScale/walder/issues/81))

## [2.0.3] - 2020-08-31

### Fixed
- Example config.yaml: cannot read variable writer (see [issue 29](https://github.com/KNowledgeOnWebScale/walder/issues/29))

lib/walder.js
'use strict';

const morgan = require('morgan');
const express = require('express');
const createLogger = require('./create-logger');
const util = require('util');
const server = require('./server');
const Path = require('path');
const pjson = require('../package');
const RouteInfo = require('./models/route-info');

// Parsers
const parseResources = require('./parsers/resource-parser');
const parseRoute = require('./parsers/route-parser');
const parseGraphQLLD = require('./parsers/query-parser');
const parseParameter = require('./parsers/parameter-parser');
const parsePipeModule = require('./parsers/pipe-module-parser');
const parseHTML = require('./parsers/html-parser');
const parseConfigFile = require('./parsers/config-file-parser');
const parseDataSources = require('./parsers/data-sources-parser');

// Loaders
const PipeModuleLoader = require('./loaders/pipe-module-loader');

// Handlers
const RequestHandler = require('./handlers/request-handler');
const GraphQLLDHandler = require('./handlers/graphql-ld-handler');
const SPARQLHandler = require('./handlers/sparql-handler');

// Validators
const Validator = require('./validators/main-validator');
const TemplateLoader = require("./loaders/template-loader");

const defaultOptions = {port: 3000, logging: 'info', cache: true, lenient: false};

module.exports = class Walder {
  constructor(configFile, options = {}) {
    if (!configFile) {
      throw Error('Configuration file is required.')
    }

    options = {...defaultOptions, ...options};

    this.version = pjson.version;
    this.port = options.port;
    this.cache = options.cache;
    this.lenient = options.lenient;
    this.app = express();
    this.logging = options.logging;
    this.configFile = configFile;

    if (!Path.isAbsolute(this.configFile)) {
      this.configFile = Path.resolve(process.cwd(), configFile);
    }

    this.cwd = Path.dirname(this.configFile);
  }

  /**
   * Starts the server.
   */
  async activate() {
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //                                                                                                                //
    //                                             Logging initialisation                                             //
    //                                                                                                                //
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    this.logger = createLogger(this.logging);

    // Set up morgan stream
    const logStream = {
      write: (message) => this.logger.info(message.replace(/\n$/, ''))
    };

    this.app.use(morgan('Received :method :url :req[accept]', { stream: logStream, immediate: true } ));
    this.app.use(morgan('Done :method :url :req[accept] :status :response-time ms - :res[content-length]', { stream: logStream } ));

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //                                                                                                                //
    //                                             Express initialisation                                             //
    //                                                                                                                //
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    this.logger.silly(`Walder v${this.version} booting up!`);

    this.app.use(express.json());
    this.app.use(express.urlencoded({extended: false}));

    await this.initialise();

    // error handler
    this.app.use(function (err, req, res, next) {
      // set locals, only providing error in development
      res.locals.message = err.message;
      res.locals.error = req.app.get('env') === 'development' ? err : {};
    });

    this.server = server.initialise(this.app, this.port, this.logger);
  };

  /**
   * Stops the server.
   */
  deactivate() {
    this.logger.info('Deactivating...');
    this.server.close((error) => {
      if (error) {
        throw Error(error);
      }
      this.logger.info('Server shut down.')
    })
  }

  /**
   * Parse and process the config file.
   */
  async initialise() {
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //                                                                                                                //
    //                                              Config file parsing                                               //
    //                                                                                                                //
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    // Parse the config file
    let yamlData
    try {
      yamlData = await parseConfigFile(this.configFile);
    } catch (err) {
      this.logger.error('Config file is invalid.');
      this.logger.error(err.message);

      throw err;
    }

    // Resources
    const resources = parseResources(yamlData['x-walder-resources'], this.cwd, this.logger);

    let sparqlHandler;

    const getSPARQLHandler = () => {
      return sparqlHandler;
    }

    // Data sources
    const graphQLLDHandler = new GraphQLLDHandler(this.logger, resources['pipe-modules'], getSPARQLHandler);
    sparqlHandler = new SPARQLHandler(this.logger, resources['pipe-modules'], graphQLLDHandler);
    const dataSources = await parseDataSources({dataSources: yamlData['x-walder-datasources'], graphQLLDHandler, pipeModulesPath: resources['pipe-modules'], cache: this.cache});


    // Parse the default error pages section
    const defaultErrorPages = parseHTML(yamlData['x-walder-errors'], resources.views, resources.layouts);

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //                                                                                                                //
    //                                            Config file validation                                              //
    //                                                                                                                //
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    this.templateLoader = new TemplateLoader();
    const validator = new Validator({templateLoader: this.templateLoader, logger:this.logger});

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //                                                                                                                //
    //                                               Start static server                                              //
    //                                                                                                                //
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    // Start static file server
    this.logger.info('Starting static server...');
    this.app.use(express.static(resources.public));
    this.logger.info('Static server running.');

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //                                                                                                                //
    //                                          Express Route initialisation                                          //
    //                                                                                                                //
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    this.requestHandler = new RequestHandler(this.logger, graphQLLDHandler, this.templateLoader, sparqlHandler);

    // Iterate over routes
    for (let path in yamlData.paths) {
      this.logger.info('Parsing route %s', path);
      for (let method in yamlData.paths[path]) {
        this.logger.info('    - Parsing method %s', method);

        // Parse required route, GraphQL-LD and postprocessing info
        this.logger.verbose('        * Parsing the express route information');
        const routeInfo = parseRoute(path, method);

        this.logger.verbose('        * Parsing the parameters description');
        const parameters = parseParameter(yamlData.paths[path][method].parameters);

        this.logger.verbose('        * Parsing the GraphQL-LD information');
        const graphQLLDInfo = parseGraphQLLD(yamlData.paths[path][method]['x-walder-query'], {defaultDataSources: dataSources, cache: this.cache, lenient: this.lenient,  parameters});

        this.logger.verbose('        * Parsing the pipe modules section');
        const pipeModules = parsePipeModule(yamlData.paths[path][method]['x-walder-postprocessing'], resources['pipe-modules']);

        this.logger.verbose('        * Parsing the html template(s)');
        const htmlInfoDictionary = parseHTML(yamlData.paths[path][method].responses, resources.views, resources.layouts);

        // Validate current path and build the cache - method
        this.logger.verbose('        * Validating the current path and method');
        await validator.validateAll({routeInfo, parameters, graphQLLDInfo, htmlInfoDictionary});

        // Complete the htmlInfoDictionary object with possible missing default error pages
        for (let statusCode in defaultErrorPages) {
          if (!(statusCode in htmlInfoDictionary)) {
            htmlInfoDictionary[statusCode] = defaultErrorPages[statusCode];
          }
        }

        // Load remote and local pipe modules
        try {
          const pipeFunctions = PipeModuleLoader.load(pipeModules);

          const callBack = this.requestHandler.handle(graphQLLDInfo, pipeFunctions, htmlInfoDictionary);

          switch (method) {
            case 'get':
              this.app.get(routeInfo.path, callBack);
              break;
            case 'post':
              this.app.post(routeInfo.path, callBack);
              break;
            case 'put':
              this.app.put(routeInfo.path, callBack);
              break;
            case 'patch':
              this.app.patch(routeInfo.path, callBack);
              break;
            case 'head':
              this.app.head(routeInfo.path, callBack);
              break;
            default:
              throw Error(util.format('"%s" is not a supported HTTP routing method.', method));
          }
        } catch (e) {
          this.logger.error(e.message);
        }
      }
    }

    // Validate default error pages
    this.logger.verbose('Validating the default error pages');
    await validator.validateAll({routeInfo: new RouteInfo('default error pages', 'any'), htmlInfoDictionary: defaultErrorPages});

    // Finish the input validation
    validator.finish();

    // The 404 Route
    this.app.get('*', this.requestHandler.handle({}, [], defaultErrorPages, 404));
  }
};
